src/widget/wtrackmenu.cpp

```diff
@@ -1528,6 +1549,127 @@ void WTrackMenu::slotClearAllMetadata() {
+
+  private:
+    void doApply(
+            const TrackPointer& pTrack) const override {
+        QString location = pTrack->getLocation();
+        qDebug() << "_";
+        qDebug() << "   RemoveTrackFilesFromDiskTrackPointerOperation";
+        qDebug() << "       " << location;
+        qDebug() << "_";
+        if (location.isNull()) {
+            return;
+        }
+        QFile file(location);
+        auto trackRef = TrackRef::fromFileInfo(
+                pTrack->getFileInfo(),
+                pTrack->getId());
+        if (!trackRef.isValid()) {
+            return;
+        }
+        QList<TrackRef> trackRefs;
+        // QList<TrackId> trackIds;
+        // TrackId id = pTrack->getId();
+        if (!file.exists()) {
+            // Missing file, purge.
+            trackRefs.append(trackRef);
+            m_pTrackCollectionManager->purgeTracks(trackRefs);
+            // m_pTrackCollectionManager->hideTracks(trackIds);
+            // trackIds.append(id);
+            return;
+        }
+        // TODO Send dialog return to purgeTracks
+        // TODO Also show reason of failure?
+        if (!file.remove()) {
+            return;
+        } else {
+            // successfully deleted, now purge.
+            trackRefs.append(trackRef);
+            m_pTrackCollectionManager->purgeTracks(trackRefs);
+            // m_pTrackCollectionManager->hideTracks(trackIds);
+            // trackIds.append(id);
+            return;
+        }
+    }
+
+    TrackCollectionManager* m_pTrackCollectionManager;
+};
+
+} // anonymous namespace
+
+void WTrackMenu::slotRemoveFromDisk() {
+    // Remove duplicates to prevent delete errors later.
+    // Update: only applies to file list in dialog,
+    // deletion is handled in independet batch operation
+    QList<TrackRef> trackRefsToDelete;
+    trackRefsToDelete.reserve(getTrackCount());
+    for (const auto& trackRef : getTrackRefs()) {
+        if (!trackRefsToDelete.contains(trackRef)) {
+            trackRefsToDelete.append(trackRef);
+        }
+    }
+
+    QList<QString> trackLocations;
+    trackLocations.reserve(trackRefsToDelete.size());
+    for (const auto& trackRef : trackRefsToDelete) {
+        DEBUG_ASSERT(trackRef.hasLocation());
+        trackLocations.append(trackRef.getLocation());
+    }
+
+    // TODO Use a dialog with a QListView that can be scrolled horizontally?
+    // Or improve linebreaks and formatting in the current plain textview.
+    // TODO Should each item may have a checkbox to allow removing it from the delete list
+    // or is it okay for the user to Cancel and adjust the selection in the tracks table?
+    // TODO Allow to keep all selected tracks' references
+
+    // Primitive hack to populate a pseudo list for the delete warning.
+    QString trackLocationsStr;
+    trackLocationsStr.append(QStringLiteral("&#8226; "))
+            .append(trackLocations.join(QStringLiteral("<br>&#8226; ")));
+    QMessageBox msgBoxDelete(QMessageBox::Critical,
```

Comment:
The dialog is unusable when selecting more than a handful of tracks. Buttons are inaccessible.

Reply:
I know...just an ugly workaround for now. thus the TODOs around here.
I need to create a custom dialog that can display a scrollable file list.

Reply:
This could be done when finalizing the PR.

First we need to get the iteration and progress reporting right. Please have a look at `TrackByIdCollectionIterator`, this should work out of the box for this purpose. All tracks are supposed to have an id. Otherwise we need to add an `TrackByRefCollectionIterator`.

Reply:
> Thanks for your feedback, appreciated.
> I'll need to wrap my head around those iterators & stuff here.

trackcollectioniterator.h

Reply:
I'll take a look.

besides:
we could let the iterator do the deletion, then use the existing ref list to check which files are not present anymore, purge those at once and report the others.

Reply:
I agree. Do the file deletion first and thereby collect two distinct sets/lists of track references: File deletion either succeeded or failed. Then purge all the successful deleted tracks at once and report the failures. You need to use `mutable` members in the TrackPointerOperation.

This procedure is eventually consistent and safe. Tracks of already deleted files will be discovered as *missing* after a rescan in case anything goes wrong.

Reply:
alrigth, have to learn about muatble first ´°_o

Reply:
The `apply()` member function is declared as `const`, because from the outside view any invocation of it must not change the observable state. You will stumble upon this ;)